### ISSUE TITLE
question - on_connect on_disconnect - ordering wrong causing reconnects.

### ISSUE THREAD

**Issue by Laro88** [eclipse/mosquitto#1069]
I am on windows using mosquittopp (c++) and having issue with manually keeping the connection up and running (the loop_start is not good on windows)

I have noticed that I get an on_connect and an immediate on_disconnect. It corresponds nicely with the verbose ooutput of mosquitto.exe -v
```
1543998217: Sending SUBACK to LJERA1
1543998217: New connection from ::1 on port 1883.
1543998217: Client LJERA1 already connected, closing old connection.
1543998217: Socket error on client LJERA1, disconnecting.
1543998217: New client connected from ::1 as LJERA1 (c1, k600).
1543998217: No will message specified.
1543998217: Sending CONNACK to LJERA1 (0, 0)
```
So I end up on a loop where the ordering triggers my feeble attempt to recover.

Is there a way to probe for the actual connection state when getting an "old" on_disconnect or another strategy for keeping a good connection - on windows...

**Comment by ralight** [comment]
```
1543998217: Client LJERA1 already connected, closing old connection.
```

This might be the key line. You are connecting again when you're already connected. Do you have a small example that shows what you are doing? It might be the easiest way to fix things.

Something like this should work though:

```
while(run){
    rc = mosquitto_connect(mosq, ...);
    if(rc){
        printf("Unable to connect: %s\n", mosquitto_strerror(rc));
        sleep()
    }else{
        do{
            rc = mosquitto_loop(mosq, -1, 1);
        } while(rc == MOSQ_ERR_SUCCESS);
        mosquitto_disconnect(mosq);
    }
}
```

**Comment by karlp** [comment]
I'm also curious why loop_start is no good _on windows_  I mean, if you don't want to deal with the threaded implications, fine, but I'm not sure I follow what about that is windows specific

**Comment by ralight** [comment]
@karlp The current windows builds provided by the project don't have threading enabled to make distribution easier.

**Comment by Laro88** [comment]
I have made a windows shell test project. There are a couple of threads, one running a kill timer, one running a messager pusher (create fake load) and one calling loop.
The primary observation that reproduces nicely (using 32 bit mosquitto.exe -v) is
mosquitto says  "Socket error on client thewrapper, disconnecting" in the log, and my test application get a code 14, then code 4 onwards.
It seems to be related to the load, however it is not reproducing nicely.
[MQTTtest.zip](https://github.com/eclipse/mosquitto/files/2649022/MQTTtest.zip)

**Comment by Laro88** [comment]
Btw. It seems that we get MOSQ_ERR_ERRNO when errno = WSAGetLastError(); is not EINTR in loop.c line 140.
Anywhere that we could dump this windows error - or make it available?

**Comment by Laro88** [comment]
I figured out part of the problem - threading. I put a couple of mutexes into my example so that there is only ONE thread accessing mosquitto.lib and stability is restored.

I can reproduce the unspecified MOSQ_ERR_ERRNO error without thread syncronization so something is probably messing up the buffers in Winsock2 or ws2tcpip - and since the documentation clearly states that the libraries are not thread safe then my bad :-)

Basically: My initial strange ordering problem was caused by my unsafe use of threading when using mosquitto - at least I can't reproduce it any more.
